### ISSUE TITLE
Bounds Inference problem for sliding over guard with if

### ISSUE THREAD

**Issue by UncleHandsome** [halide/Halide#6044]
Hello
Here is my algorithm
`output(x, y) = input(x, y) + input(x, y + 1) + input(x, y + 2) + input(x, y + 3) + input(x, y + 4);
`

I tried the following schedule in Generator

    add_requirement(output.dim(0).extent() > 0);    
    add_requirement(output.dim(1).extent() > 0);   
  
    output.compute_root()      
      .bound(x, 0, output.width())    
      .bound(y, 0, output.height())  
      .split(y, y, yo, 8, TailStrategy::GuardWithIf)  
      .split(yo, yo, yoo, 2, TailStrategy::GuardWithIf)  
      .unroll(yo)
      .unroll(yoo)
      ;

    input.in()      
      .compute_at(output, yo)   
      .store_at(output, y)   
      .unroll(_1)   
      ;  
  

In this case, Halide will infer the allocation bounds of input.in() to 8.
However, if I tried the following schedule,
Halide won't be able to infer the bounds of input.in() to a constant value.
In this case, I think the correct answer should be 8 * 2 = 16.
Thank you

    add_requirement(output.dim(0).extent() > 0);
    add_requirement(output.dim(1).extent() > 0);
    output.compute_root()
        .bound(x, 0, output.width())
        .bound(y, 0, output.height())
        .split(x, x, vx, 2, TailStrategy::RoundUp)
        .split(y, y, yo, 8, TailStrategy::GuardWithIf)
        .split(yo, yo, yoo, 2, TailStrategy::GuardWithIf)
        .unroll(yo)
        .unroll(yoo)
        .vectorize(vx)
        .reorder(vx, yoo, yo, y, x)
        ;
    input.in()
      .compute_at(output, yo)
      .store_at(output, y)
      .unroll(_1)
      .vectorize(_0)
      ;

**Comment by UncleHandsome** [comment]
Below if the source code I used to run this example.
Thank you!

    #include "Halide.h"
    #include "halide_benchmark.h"
    #include <cstdio>
    #include <memory>

    using namespace Halide;
    using namespace Halide::Tools;

    class Example : public Generator<Example> {
    public:
        using Generator<Example>::Generator;

        Input<Buffer<uint8_t>> input{"input", 2};
        Output<Buffer<uint8_t>> output{"output", 2};

        void generate() {
            output(x, y) = input(x, y) + input(x, y + 1) + input(x, y + 2) + input(x, y + 3) + input(x, y + 4);
        }

        void schedule() {
            output.compute_root()
                .bound(x, 0, output.width())
                .bound(y, 0, output.height())
                .split(x, x, vx, 2, TailStrategy::RoundUp)
                .split(y, y, yo, 8, TailStrategy::GuardWithIf)
                .split(yo, yo, yoo, 2, TailStrategy::GuardWithIf)
                .unroll(yo)
                .unroll(yoo)
                .vectorize(vx)
                .reorder(vx, yoo, yo, y, x)
                ;

            add_requirement(output.dim(0).extent() > 0);
            add_requirement(output.dim(1).extent() > 0);

            input.in()
                .compute_at(output, yo)
                .store_at(output, y)
                .unroll(_1)
                .vectorize(_0)
                ;
        }

    private:
        Var x{"x"}, y{"y"}, vx{"vx"}, yo{"yo"}, xi{"xi"}, xii{"xii"}, yoo{"yoo"};
    };


    int main(int argc, char **argv) {
        GeneratorContext context(get_jit_target_from_environment());
        Buffer<uint8_t> input(100, 104);
        auto gen = context.apply<Example>(input);

        Buffer<uint8_t> img = gen->realize({100, 100});

        printf("Success!\n");
        return 0;
    }

**Comment by rootjalex** [comment]
What version/commit of Halide are you using? I actually see the opposite behavior (on current master) - the first piece of code does not produce a constant allocation:
```
allocate input_im_global_wrapper$0[uint8 * output.extent.0 * (max(min(input_im_global_wrapper$0._1.extent_realized.s, 8), -4) + 4)]
```
While the second schedule does produce a constant-sized allocation:
```
allocate input_im_global_wrapper$0[uint8 * 24]
```

**Comment by UncleHandsome** [comment]
Hello @rootjalex 
Thank you for your help!
I tried latest commit 677bf30671 and found both version of the code will fail in UnrollLoops.
I have updated UnrollLoops to unroll from outer to inner and it will run successfully.
The allocation can be 24, but the index doesn't simplify away by Halide.
Below is my change to Unroll Loops.

    Comment out Line 76
    Change line 84 to substitute(for_loop->name, for_loop->min + i, body)

Is it possible for Halide to simplify this kind of split?
Thank you

**Comment by rootjalex** [comment]
Hi @UncleHandsome , sorry for the delay. UnrollLoops was getting a nasty expression for an extent that is hard for interval arithmetic to analyze:
```
((min((((output.s0.y.y*4) + output.s0.y.yo.yo)*2) + 2, output.extent.1) - select(0 < output.s0.y.yo.yo, min(((output.s0.y.y*4) + output.s0.y.yo.yo)*2, output.extent.1) + 4, ((output.s0.y.y*4) + output.s0.y.yo.yo)*2)) + 4)
```
This expression is upper bounded by 6, but it's difficult to discover that without some very specific rewrites being performed.
I've added a fix in #6059 .

**Comment by UncleHandsome** [comment]
Hello @rootjalex 
Thank for your help!
This kind of usage is very tricky.
I'll try it after it is stabilized!
Thank you!
